### PR TITLE
fix: getKebabCase #294

### LINKED ISSUE #294: Fix `getKebabCase()` implementation
## Description

Our current `getKebabCase()` implementation is not perfect, there are some edge cases that we have not covered properly.

## Expected Behaviour

| String | Actual | Expected |
|-------|------- |----------|
| DKI Jakarta | dkijakarta | dki-jakarta |
| Daftar Kontak Fasilitas & Alat Kesehatan per Provinsi | daftar-kontak-fasilitas-alat-kesehatanper-provinsi | daftar-kontak-fasilitas-alat-kesehatan-per-provinsi |
| Laman Edukasi COVID-19 | laman-edukasi-covid19 | laman-edukasi-covid-19 |
| Donasi dan Penggalangan Dana | donasidan-penggalangan-dana | donasi-dan-penggalangan-dana |
| D.I. Yogyakarta | d.i.yogyakarta | d-i-yogyakarta |
| Ambulancesiaga.com | ambulancesiaga.com | ambulancesiaga-com |
| "PUTRA SOERADI" Jual Peti Mati | putrasoeradijual-peti-mati | putra-soeradi-jual-peti-mati |
| Hoo Hap Hwee (Perkumpulan Budi Abadi) | hoo-hap-hwee-perkumpulan-budi-abadi | hoo-hap-hwee-perkumpulan-budi-abadi |
| UD. Bpk. Giyanto Sedia/Jual Peti Jenazah | ud.bpk.giyanto-sedia-jual-peti-jenazah | ud-bpk-giyanto-sedia-jual-peti-jenazah |
| Samator, PT. Aneka Gas Industri, Tbk. | samator-pt.aneka-gas-industri-tbk. | samator-pt-aneka-gas-industri-tbk |
## Implementation Model

Before we start making the changes, we need to properly setup the unit test infrastructure #178 and cover the function with extensive test cases.

### PR DESCRIPTION
**Pull request by fncolon**
Fixes #294  #110 

## Description

Convert the whitespace and underscore to minus sign without replace the spaces with camel case to fix the issue.
```
replace(/[\s_]+/g, '-')
```

## Current Tasks

- [x] Fix: getKebabCase

### CODE REVIEW

**Comment by fncolon** [comment]
PoC
Capture

File: lib/string-utils.ts [hunk 1]
```diff
@@ -71,9 +71,7 @@ export function convertToKebabCase(str: string): string {
  */
 export function getKebabCase(str?: string): string {
   return convertToKebabCase(
-    removeSpaces(
-      replaceSpecialCharacterWithSpace(replaceSpacesWithCamelCase(str ?? "")),
-    ),
+    replaceSpecialCharacterWithSpace(str),
```
Comment (ekamuktia) [inline lib/string-utils.ts:74]: Please change this to 
`replaceSpecialCharacterWithSpace(str ?? "")`

because `str` can be undefined type

**Review by ekamuktia — CHANGES_REQUESTED** [review]
For the test case mentioned in #294 it's okay

But there are some cases that still not covered, for example:
- Text containing special character
"PUTRA SOERADI" Jual Peti Mati: -putra-soeradi-jual-peti-mati
Hoo Hap Hwee (Perkumpulan Budi Abadi): hoo-hap-hwee-perkumpulan-budi-abadi-
Please trim trailing `-`

Also for text containing `.`, what is the rule? 
Could you confirm the rule and include it on test case? @zainfathoni 
For example, with updated function, it'll be:
D.I. Yogyakarta: d.i.-yogyakarta
UD. Bpk. Giyanto Sedia/Jual Peti Jenazah: ud.-bpk.-giyanto-sedia-jual-peti-jenazah
Ambulancesiaga.com: ambulancesiaga.com
Samator, PT. Aneka Gas Industri, Tbk.: samator-pt.-aneka-gas-industri-tbk.
I think the `.` should be removed, or replaced with `-`

File: lib/string-utils.ts (hunk 1 above)
Comment (fncolon) [reply lib/string-utils.ts:74]: will do , thanks

**Comment by zainfathoni** [comment]
Please add all the edge cases for into the issue description, Mbak @ekamuktia. We can discuss all the possibilities there.

You should have had access to edit the issue description because I have invited you as a maintainer of this repository. If you aren't able to do it, please check your email to accept the invitation, thanks!

**Comment by ekamuktia** [comment]
Test cases updated in #294 description

**Comment by fncolon** [comment]
> Test cases updated in #294 description

Capture

just done it with the new test cases, will start commiting ASAP

**Review by ekamuktia — CHANGES_REQUESTED** [review]
Please fix parts that causing eslint error by adding new line as suggested on the error message

https://github.com/kawalcovid19/wargabantuwarga.com/pull/313/checks?check_run_id=3140847528

**Comment by zainfathoni** [comment]
Thanks for working on it @fncolon, but we're not merging until we set up a proper integration test for #178, I'm currently working on it at #358. :pray:

**Comment by fncolon** [comment]
> Thanks for working on it @fncolon, but we're not merging until we set up a proper unit & integration tests infrastructure for #178, I'm currently working on it at #358. 🙏

Thanks sir Zain, i will keep watching this PR to avoid merge conflicts next time, thanks 🙏

**Comment by zainfathoni** [comment]
I've merged #358, so please continue with this PR while updating the test cases to match the expected values from the descriptions. Thanks, @fncolon! 🙏

**Comment by fncolon** [comment]
> I've merged #358, so please continue with this PR while updating the test cases to match the expected values from the descriptions. Thanks, @fncolon! 🙏

Hello sir Zain, the expected value from the issue and the test is different, which one should i follow?

**Comment by zainfathoni** [comment]
> > I've merged #358, so please continue with this PR while updating the test cases to match the expected values from the descriptions. Thanks, @fncolon! 🙏
> 
> Hello sir Zain, the expected value from the issue and the test is different, which one should i follow?

That expected value was for the old implementation, we need to put those wrong values so that the test doesn't break the `main` branch build. You should update the test to the actual result of your updated implementation, if the tests are successful, it proves that your regex actually works. 😉

**Comment by fncolon** [comment]
> > > I've merged #358, so please continue with this PR while updating the test cases to match the expected values from the descriptions. Thanks, @fncolon! 🙏
> > 
> > 
> > Hello sir Zain, the expected value from the issue and the test is different, which one should i follow?
> 
> That expected value was for the old implementation, we need to put those wrong values so that the test doesn't break the `main` branch build. You should update the test to the actual result of your updated implementation, if the tests are successful, it proves that your regex actually works. 😉

already update the unit test based on current implementation

File: lib/string-utils.ts [hunk 2]
```diff
@@ -70,11 +73,7 @@ export function convertToKebabCase(str: string): string {
  * @returns {string} kebab-cased version of `str`
  */
 export function getKebabCase(str?: string): string {
-  return convertToKebabCase(
-    removeSpaces(
-      replaceSpecialCharacterWithSpace(replaceSpacesWithCamelCase(str ?? "")),
-    ),
-  );
+  return convertToKebabCase(replaceSpecialCharacterWithSpace(str ?? "").trim());
```
Comment (zainfathoni) [inline lib/string-utils.ts:76]: Unfortunately, the [Unit and integration tests CI check](https://github.com/kawalcovid19/wargabantuwarga.com/pull/313/checks?check_run_id=3151271445#step:4:111) still failed due to the lower code coverage.
Could you please write more tests to cover the rest of the functions in this `string-utils.ts` file?
Thanks, @fncolon! 🙏

**Review by zainfathoni — CHANGES_REQUESTED** [review]

File: lib/string-utils.ts (hunk 2 above)
Comment (fncolon) [reply lib/string-utils.ts:76]: Oh i see, will write more tests, after take a sleep for few hours. 

Thanks.

Comment (zainfathoni) [reply lib/string-utils.ts:76]: Thanks, have a nice rest, @fncolon, take care of yourself! 😄

Comment (zainfathoni) [reply lib/string-utils.ts:76]: I just added some tests on 41ddae9 to increase the code coverage just enough to meet the threshold. 😁

**Review by zainfathoni — APPROVED** [review]
LGTM 💯 

Based on the exhaustive test on `getKebabCase`, now we're pretty confident that the function works as expected. Thanks a lot, @fncolon! 🙏

**Pull request merged**